### PR TITLE
Fix falsely big bit strings on the instances with the test data (neg. patient identifiers)

### PR DESCRIPTION
**Pull request by forus**
Given that we used to start patient identifiers with 1 000 384 597, the negative patient identifiers suggested huge bitset strings.
That caused the bitset patient counts to become unreasonably slow.

Negative patient identifiers are still uploaded with "make postgres_test". The change to start the patient sequence come to rescue here. Ticket to load test-studies with transmart-copy will get rid off negative patient ids completely https://jira.thehyve.nl/browse/TMT-242

### CODE REVIEW

**Comment by forus** [comment]
While working on current changes, the limitation of current patient mapping was realized. See https://jira.thehyve.nl/browse/TMT-249

File: transmart-data/updatedb/release-17.1-HYVE-4/postgres/fix_neg_patient_nums.sql [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		alter table i2b2demodata.relation alter constraint right_subject_id_fk deferrable initially immediate;
+		set constraints i2b2demodata.left_subject_id_fk deferred;
+		set constraints i2b2demodata.right_subject_id_fk deferred;
+		IF min_pos_patient_num IS NULL THEN
+			RAISE NOTICE 'No patients with positive identifiers were found. Using the patient id sequence instead.';
+			select nextval('i2b2demodata.SEQ_PATIENT_NUM') into min_pos_patient_num;
+			max_pos_patient_num := min_pos_patient_num;
+		END IF;
+		IF min_pos_patient_num > ABS(min_neg_patient_num) THEN
+			update i2b2demodata.patient_dimension set patient_num = min_pos_patient_num + patient_num where patient_num < 0;
```
Comment (gijskant) [inline transmart-data/updatedb/release-17.1-HYVE-4/postgres/fix_neg_patient_nums.sql]: Impressive function! It seems that a number of tables is missing, but these are not used by us:
- `i2b2demodata.qt_patient_enc_collection`
- `i2b2demodata.archive_observation_fact`
- `deapp.de_snp_copy_number`
- `deapp.de_snp_subject_sorted_def`
- `deapp.de_snp_data_by_patient`
- `deapp.de_subject_snp_dataset`
- `deapp/de_snp_calls_by_gsm`
and some in `tm_cz`, `tm_wz`.
We could add lines here or warn the user that the migration may not cover all tables.

Comment (gijskant) [reply transmart-data/updatedb/release-17.1-HYVE-4/postgres/fix_neg_patient_nums.sql]: Above comments lists occurences of `patient_num`. For `patient_id`, there is also the `deapp.de_subject_*_data` tables.

Comment (gijskant) [reply transmart-data/updatedb/release-17.1-HYVE-4/postgres/fix_neg_patient_nums.sql]: and `i2b2demodata.qt_patient_sample_collection`.

Comment (gijskant) [reply transmart-data/updatedb/release-17.1-HYVE-4/postgres/fix_neg_patient_nums.sql]: `i2b2demodata.encounter_mapping`.

File: transmart-data/updatedb/release-17.1-HYVE-4/README.md [hunk 2]
```diff
@@ -13,6 +13,15 @@ Views
...
 and bunch of views to count number of patients per study and concept.
 
+## Data migrations
+
+### Get rid of negative patient identifiers
+
+Given that we used to start patient identifiers with billion smth. negative patient identifiers suggest huge bitset strings.
+That cases the bitset patient counts to become unreasonably sloow.
+`fix_neg_patient_nums.sql` fixes this issue by giving new, positive ids to patients with negative ids.
+**Note:** The script fixes patient numbers in the tables used by the test studies (`test_data` folder). It does not scan other tables.
```
Comment (gijskant) [inline transmart-data/updatedb/release-17.1-HYVE-4/README.md]: Maybe include the :warning: symbol?

File: transmart-data/updatedb/release-17.1-HYVE-4/README.md [hunk 3]
```diff
@@ -13,6 +13,15 @@ Views
 PostgreSQL extension function (`pg_bitcount(bitset)`) to count number of bits in the bitset optimally,
 and bunch of views to count number of patients per study and concept.
 
+## Data migrations
+
+### Get rid of negative patient identifiers
+
+Given that we used to start patient identifiers with billion smth. negative patient identifiers suggest huge bitset strings.
+That cases the bitset patient counts to become unreasonably sloow.
```
Comment (gijskant) [inline transmart-data/updatedb/release-17.1-HYVE-4/README.md]: Two typos.

File: transmart-data/updatedb/release-17.1-HYVE-4/postgres/migrate.sql [hunk 4]
```diff
@@ -3,4 +3,5 @@
 \i ../../../ddl/postgres/biomart_user/views/patient_num_boundaries.sql
 \i ../../../ddl/postgres/biomart_user/materialized_views/study_concept_bitset.sql
 \i ../../../ddl/postgres/biomart_user/views/patient_set_bitset.sql
-\i ../../../ddl/postgres/biomart_user/views/study_concept_patient_set_bitset.sql
\ No newline at end of file
+\i ../../../ddl/postgres/biomart_user/views/study_concept_patient_set_bitset.sql
+\i fix_neg_patient_nums.sql
```
Comment (gijskant) [inline transmart-data/updatedb/release-17.1-HYVE-4/postgres/migrate.sql]: Please exclude from the migrate script, but do document the script.
This does not have to be applied as mandatory migration step for production instances.

**Pull request merged**